### PR TITLE
Avro-3767 [Rust] fix complex Ref resolving in Union

### PR DESCRIPTION
**Pull request by ftelnov**
Associated issue: https://issues.apache.org/jira/browse/AVRO-3767

### CODE REVIEW

**Comment by ftelnov** [comment]
Now the fix is ready - might contain some dirty code, please review patiently.

What was the issue: at the resolve_union method, we prematurely resolved all references in the Union. It was done in the wrong way - it was resolving only top-level schemas, without recursive lookup. I think it is bad solution even if it might work after fix - firstly, it adds complexity, secondly it is inconsistent with union validation, which does not do premature Ref resolving.

I suppose, it was done that way because `ResolvedSchema` does not handle external references. I've tried to fix it - I added a possibility to resolve external references to `ResolvedSchema`(with an extra argument and new constructor). And I adjusted `find_schema` method of the union to include external refs handle.

**Comment by ftelnov** [comment]
I also allowed myself to do a slight renaming and signature changes in ResolvedSchema - as previously it was a bit inconvenient to use and read

File: lang/rust/avro/src/schema.rs [hunk 1]
```diff
@@ -434,21 +455,28 @@ impl<'s> ResolvedSchema<'s> {
...
                 Schema::Ref { name } => {
                     let fully_qualified_name = name.fully_qualified_name(enclosing_namespace);
-                    if names_ref.get(&fully_qualified_name).is_none() {
+                    // first search for reference in current schema, then look into external references.
+                    let is_resolved_locally = self.names_ref.get(&fully_qualified_name).is_some();
+                    let is_resolved_with_known_schemas = known_schemas
+                        .as_ref()
+                        .map(|schemas| schemas.get(&fully_qualified_name).is_some())
+                        .unwrap_or(false);
+                    if !is_resolved_locally && !is_resolved_with_known_schemas {
```
[the remark this answers is not in the document]
Comment (ftelnov) [reply lang/rust/avro/src/schema.rs:479]: I hope compiler would solve such a thing for us - mostly I wrote it to enhance readability. But yeah, you are right, we'd better state it explicitly

**Comment by martin-g** [comment]
Please review https://github.com/apache/avro/pull/2266/commits/59cb2b2ca6149c102fa08bc112dc87d259248baa and let me know if I broke something!

**Comment by ftelnov** [comment]
@martin-g everything seems fine to me, despite that naming I've mentioned earlier. Good job, thanks for prettifying it!

**Comment by martin-g** [comment]
Thank you, @ftelnov !

**Pull request merged**